lib/spack/spack/solver/concretize.lp

```diff
@@ -516,6 +516,13 @@ attr("virtual_node", VirtualNode) :- attr("virtual_root", VirtualNode).
 % then the provider is the root package.
 attr("root", PackageNode) :- attr("virtual_root", VirtualNode), provider(PackageNode, VirtualNode).
 
+attr("virtual_root", VirtualNode) :- attr("root", PackageNode), provider(PackageNode, VirtualNode).
+
+% This is required for require_cflags to pass with `require: w cflags="-O1"`
+1 { attr("virtual_node", node(0..X-1, Virtual)) : max_dupes(Virtual, X) } :-
+  virtual_condition_holds(node(ProviderID, Provider), Virtual),
+  attr("node", node(ProviderID, Provider)).
```

Comment:
This will not blend with the semantic we want to merge in #35322. If a condition to provide a virtual holds, it is not necessary for the node to provide the virtual.

Reply:
Two points:

* This only says that if X provides a virtual Y, that Y must exist as a virtual. It does not say that X must be the provider for Y. Therefore I don't think it actually conflicts (at least with the high-level goal stated in #35322)
* Without this, note that  `packages:<virtual>:require` cannot guarantee imposing requirements on providers

```
packages:
  mpi:
    require: openmpi +shared
```

i.e. this can only place constraints on MPI right now when MPI is a (virtual) node in the DAG. Another option is to take constraints like `require: openmpi+shared` and generate a secondary requirement for the provider (whether or not it is providing something).

Reply:
> This only says that if X provides a virtual Y, that Y must exist as a virtual.

I don't read it exactly like that, but rather: _if X satisfies the conditions to provide the virtual Y, then Y must exist as a virtual_. This is not always true. For instance, if in a solve `intel-parallel-studio` is used only for `mpi` there does not need to be a `lapack` virtual node in the answer set.

Reply:
> I don't read it exactly like that, but rather: if X satisfies the conditions to provide the virtual Y, then Y must exist as a virtual.

Agree.

> if in a solve intel-parallel-studio is used only for mpi there does not need to be a lapack virtual node in the answer set.

What do you think should happen for something like:

```
packages:
  lapack:
    require:
    - one_of: [openblas, netlib-lapack]

...

spack spec intel-parallel-studio+mkl
```

? 

(I don't think this is a counter example to your claim, but I am curious about the desired behavior).

Overall, it sounds like you'd prefer the second strategy I described in https://github.com/spack/spack/pull/39880#discussion_r1362499897

Reply:
> What do you think should happen for something like: ...

I think it should install, since we asked for the `intel-parallel-studio+mkl` package explicitly.